kong/pdk/response.lua

```diff
@@ -517,14 +519,94 @@ local function new(self, major_version)
   end
 
 
-  --function _RESPONSE.set_raw_body(body)
-  --  -- TODO: implement, but how?
-  --end
+  ---
+  -- Returns the full body when the last chunk has been read
+  --
+  -- @function kong.response.get_raw_body
+  -- @phases `body_filter`
+  -- @treturn string body The full body when the last chunk has been read,
+  --                      otherwise returns `nil`
+  -- @usage
+  -- local body = kong.response.get_raw_body()
+  -- if body then
+  --   ...
+  -- end
+  function _RESPONSE.get_raw_body()
+    check_phase(PHASES.body_filter)
+
+    local body_buffer
+    local chunk = arg[1]
+    local eof = arg[2]
+    if eof then
+      body_buffer = self.ctx.core.body_buffer
+      if not body_buffer then
+        return chunk
+      end
+    end
+
+    if type(chunk) == "string" and chunk ~= "" then
+      if not eof then
+        body_buffer = self.ctx.core.body_buffer
+      end
```

Comment:
I found the assignment order of `body_buffer` a little confusing, perhaps it can be simplified?

```suggestion
    local body_buffer = self.ctx.core.body_buffer
    local chunk, eof = arg[1], arg[2]
    
    if eof and not body_buffer then
      -- the entire body is in this first chunk, no need to buffer
      return chunk
    end

    -- add chunk to buffer
    if type(chunk) == "string" and chunk ~= "" then
```

(untested)

Reply:
it was done to minimize context reads